### PR TITLE
Dev interp inlet

### PR DESCRIPTION
**Pull request by shaering**
Adds base class for handling external data and gaussian interpolation of such data for an inlet bc.  Gaussian stencil based on points within radius defined by second-nearest data point to each dof.  Currently only supports csv files of the format:
no, x, y, z, temp, u, v, w

### CODE REVIEW

File: src/calorically_perfect.cpp [hunk 1]
```diff
@@ -283,28 +288,58 @@ void CaloricallyPerfectThermoChem::initializeSelf() {
...
         if (rank0_) {
-          std::cout << "ERROR: Inlet type = " << type << " not supported." << std::endl;
+          std::cout << "Calorically Perfect: Setting interpolated Dirichlet temperature on patch = " << patch
+                    << std::endl;
         }
-        assert(false);
-        exit(1);
+        AddTempDirichletBC(temperature_field_, inlet_attr);
+
+        // copy interpolated bc onto initial field
```
[the remark this answers is not in the document]
Comment (shaering) [reply src/calorically_perfect.cpp:313]: Replaced with 
Tn_gf_.ProjectBdrCoefficient(*temperature_field_, inlet_attr);

File: src/calorically_perfect.hpp [hunk 2]
```diff
@@ -128,6 +130,9 @@ class CaloricallyPerfectThermoChem : public ThermoChemModelBase {
   ParGridFunction R0PM0_gf_;
   ParGridFunction Qt_gf_;
 
+  // ParGridFunction *buffer_tInlet_ = nullptr;
+  GridFunctionCoefficient *temperature_field_ = nullptr;
```
[the remark this answers is not in the document]
Comment (shaering) [reply src/calorically_perfect.hpp:133]: Renamed.

Regarding the second point (and the updateBC comment below) don't we need to keep the field to modify with ramping/time-dependant bc's?

File: src/calorically_perfect.cpp [hunk 3]
```diff
@@ -283,28 +288,58 @@ void CaloricallyPerfectThermoChem::initializeSelf() {
...
         }
-        assert(false);
-        exit(1);
+        AddTempDirichletBC(temperature_value, inlet_attr);
+
       } else if (type == "interpolate") {
+        Array<int> inlet_attr(pmesh_->bdr_attributes.Max());
+        inlet_attr = 0;
+        inlet_attr[patch - 1] = 1;
+        temperature_field_ = new GridFunctionCoefficient(extData_interface_->Tdata);
```
[the remark this answers is not in the document]
Comment (shaering) [reply src/calorically_perfect.cpp:306]: Comment about modifying bc's later in the simualtion.

File: src/tomboulides.cpp [hunk 4]
```diff
@@ -185,6 +233,8 @@ void Tomboulides::initializeSelf() {
   // Initialize minimal state and interface
   vfec_ = new H1_FECollection(vorder_, dim_);
   vfes_ = new ParFiniteElementSpace(pmesh_, vfec_, dim_);
+  sfec_ = new H1_FECollection(vorder_, dim_);
```
[the remark this answers is not in the document]
Comment (shaering) [reply src/tomboulides.cpp:260]: Will there every be a case where we will not use equal order for pressure in Tomb?  If not, you are correct.  At any rate, deleted for now.

File: src/tomboulides.hpp [hunk 5]
```diff
@@ -104,6 +104,9 @@ class Tomboulides final : public FlowBase {
   const int porder_;
   const int dim_;
 
+  // number of scalar dofs per rank
+  int Sdof_;
```
[the remark this answers is not in the document]
Comment (shaering) [reply src/tomboulides.hpp:109]: done

File: src/externalData_base.hpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  extDataToFlow toFlow_interface_;
+
+  /// Interface object, provides fields necessary for the turbModel
+  extDataToThermoChem toThermoChem_interface_;
+};
+
+/**
+ * Provides trivial field
+ */
+class ZeroExternalData final : public ExternalDataBase {
```
[the remark this answers is not in the document]
Comment (shaering) [reply src/externalData_base.hpp:94]: Commented in slack.  Delete if you do not want.

File: src/gaussianInterpExtData.cpp [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+        Udata[n + 1 * Sdof_] = 0.0;
+        Udata[n + 2 * Sdof_] = 0.0;
+        Tdata[n] = 0.0;
+        // std::cout << " iCount of zero..." << endl; fflush(stdout);
+      }
+    }
+    //}
+  }
+
+  /*
```
[the remark this answers is not in the document]
Comment (shaering) [reply src/gaussianInterpExtData.cpp:390]: See comment in slack.  Fine to delete if concern is not valid.

File: src/tomboulides.hpp [hunk 8]
```diff
@@ -136,13 +139,19 @@ class Tomboulides final : public FlowBase {
...
 
-  /// Presser FEM objects and fields
+  /// Pressure FEM objects and fields
   mfem::FiniteElementCollection *pfec_ = nullptr;
   mfem::ParFiniteElementSpace *pfes_ = nullptr;
   mfem::ParGridFunction *p_gf_ = nullptr;
   mfem::ParGridFunction *resp_gf_ = nullptr;
 
+  /// Scalar FEM objects of vorder
+  mfem::FiniteElementCollection *sfec_ = nullptr;
```
[the remark this answers is not in the document]
Comment (shaering) [reply src/tomboulides.hpp:154]: removed.

File: src/tomboulides.cpp [hunk 9]
```diff
@@ -98,9 +99,54 @@ Tomboulides::Tomboulides(mfem::ParMesh *pmesh, int vorder, int porder, temporalS
...
+        if (pmesh_->GetMyRank() == 0) {
+          std::cout << "Tomboulides: Setting uniform Dirichlet velocity on patch = " << patch << std::endl;
+        }
+        addVelDirichletBC(velocity_value, inlet_attr);
+
+      } else if (type == "interpolate") {
+        Array<int> inlet_attr(pmesh_->bdr_attributes.Max());
+        inlet_attr = 0;
+        inlet_attr[patch - 1] = 1;
+        velocity_field_ = new VectorGridFunctionCoefficient(extData_interface_->Udata);
```
[the remark this answers is not in the document]
Comment (shaering) [reply src/tomboulides.cpp:141]: See comments about time-dependant bc's in temp

File: src/calorically_perfect.cpp [hunk 10]
```diff
@@ -156,6 +156,8 @@ void CaloricallyPerfectThermoChem::initializeSelf() {
   sfec_ = new H1_FECollection(order_);
   sfes_ = new ParFiniteElementSpace(pmesh_, sfec_);
 
+  Sdof_ = sfes_->GetNDofs();
```
[the remark this answers is not in the document]
Comment (shaering) [reply src/calorically_perfect.cpp:161]: deleted

File: src/calorically_perfect.hpp [hunk 11]
```diff
@@ -116,6 +116,8 @@ class CaloricallyPerfectThermoChem : public ThermoChemModelBase {
   // Scalar \f$H^1\f$ finite element space.
   ParFiniteElementSpace *sfes_ = nullptr;
 
+  int Sdof_;
```
[the remark this answers is not in the document]
Comment (shaering) [reply src/calorically_perfect.hpp:119]: deleted

File: src/gaussianInterpExtData.cpp [hunk 12]
```diff
@@ -0,0 +1,495 @@
...
+    if (type == "interpolate") {
+      isInterpInlet_ = true;
+      tpsP_->getInput((basepath + "/name").c_str(), fname_, std::string("inletPlane.csv"));
+    } else {
+      if (rank0_) {std::cout << "Only one interpolated inlet currently supported." << endl;}
+      exit(1);
+    }
+  }
+
+  // TODO(swh): add checks and error msg for invalid sponge settings
```
[the remark this answers is not in the document]
Comment (shaering) [reply src/gaussianInterpExtData.cpp:80]: deleted

File: src/calorically_perfect.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (trevilo) [reply src/calorically_perfect.hpp]: Ok, we can leave it.  I think we can handle time dependent through the `Coefficient` object without keeping this pointer, but we can worry about that when we get there.

File: src/gaussianInterpExtData.cpp [hunk 13]
```diff
@@ -0,0 +1,495 @@
...
+    return;
+  }
+
+  rank_ = pmesh_->GetMyRank();
+  rank0_ = false;
+  if (rank_ == 0) {
+    rank0_ = true;
+  }
+  dim_ = pmesh_->Dimension();
+  nvel_ = dim_;
```
[the remark this answers is not in the document]
Comment (shaering) [reply src/gaussianInterpExtData.cpp]: removed

**Pull request merged**